### PR TITLE
Authenticated WebFinger

### PR DESCRIPTION
**Pull request by abelgardep**
## Related Issues
App: https://github.com/owncloud/android/issues/3943

Library PR (if needed): https://github.com/owncloud/android-library/pull/545

- [x] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

### CODE REVIEW

File: owncloudData/src/main/java/com/owncloud/android/data/oauth/repository/OCOAuthRepository.kt [hunk 1]
```diff
@@ -19,18 +19,41 @@
...
 import com.owncloud.android.domain.authentication.oauth.model.TokenRequest
 import com.owncloud.android.domain.authentication.oauth.model.TokenResponse
+import com.owncloud.android.domain.exceptions.FileNotFoundException
+import com.owncloud.android.domain.webfinger.model.WebfingerRel
+import timber.log.Timber
 
 class OCOAuthRepository(
-    private val oidcRemoteOAuthDataSource: RemoteOAuthDataSource
+    private val oidcRemoteOAuthDataSource: RemoteOAuthDataSource,
+    private val webfingerRemoteDatasource: WebfingerRemoteDatasource,
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/oauth/repository/OCOAuthRepository.kt:35]: To keep consistency with the rest of datasources, maybe this class should be called `RemoteWebfingerDataSource` and thus this property `remoteWebfingerDataSource`, do you agree?

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCWebfingerRemoteDatasource.kt [hunk 2]
```diff
@@ -20,19 +20,51 @@ package com.owncloud.android.data.webfinger.datasources.implementation
 import com.owncloud.android.data.ClientManager
 import com.owncloud.android.data.executeRemoteOperation
 import com.owncloud.android.data.webfinger.datasources.WebfingerRemoteDatasource
+import com.owncloud.android.domain.webfinger.model.WebfingerRel
+import com.owncloud.android.lib.common.OwnCloudClient
+import com.owncloud.android.lib.common.authentication.OwnCloudCredentialsFactory
 import com.owncloud.android.lib.resources.webfinger.services.WebfingerService
 
 class OCWebfingerRemoteDatasource(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCWebfingerRemoteDatasource.kt:28]: If you finally change the name of the interface, rename this consequently as well 😁

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/repository/OCWebfingerRepository.kt [hunk 3]
```diff
@@ -19,11 +19,33 @@ package com.owncloud.android.data.webfinger.repository
...
 import com.owncloud.android.domain.webfinger.WebfingerRepository
+import com.owncloud.android.domain.webfinger.model.WebfingerRel
 
 class OCWebfingerRepository(
     private val webfingerRemoteDatasource: WebfingerRemoteDatasource,
 ) : WebfingerRepository {
 
-    override fun getJRDFromWebFingerHost(server: String, resource: String): String =
-        webfingerRemoteDatasource.getJRDFromWebFingerHost(lookupServer = server, username = resource)
+    override fun getInstancesFromWebFinger(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/repository/OCWebfingerRepository.kt:28]: I see in some methods or classes you write `WebFinger` and in others `Webfinger`... This is not important at all but I'd try to unify it to be consistent 👍

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Good job @abelgardep! Here I left some suggestions 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/AuthenticationViewModel.kt [hunk 4]
```diff
@@ -48,32 +49,33 @@ class AuthenticationViewModel(
...
-            useCase = getJRDFromWebfingerHostUseCase,
-            useCaseParams = GetJRDFromWebfingerHostUseCase.Params(server = webfingerLookupServer, resource = webfingerUsername)
+            liveData = _legacyWebfingerHost,
+            useCase = getOwnCloudInstanceFromWebFingerUseCase,
+            useCaseParams = GetOwnCloudInstanceFromWebFingerUseCase.Params(server = webfingerLookupServer, resource = webfingerUsername)
         )
     }
 
+    private val _serverInfo = MediatorLiveData<Event<UIResult<ServerInfo>>>()
+    val serverInfo: LiveData<Event<UIResult<ServerInfo>>> = _serverInfo
```
[the remark this answers is not in the document]
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/AuthenticationViewModel.kt:77]: In this case, the livedata is next to the usecase call that will update it.

```
livedata
usecase

livedata
usecase

livedata
usecase
```
So, in this case, I would say that it is not bad at all

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import com.owncloud.android.data.ClientManager
+import com.owncloud.android.data.executeRemoteOperation
+import com.owncloud.android.data.webfinger.datasources.RemoteWebFingerDatasource
+import com.owncloud.android.domain.webfinger.model.WebFingerRel
+import com.owncloud.android.lib.common.OwnCloudClient
+import com.owncloud.android.lib.common.authentication.OwnCloudCredentialsFactory
+import com.owncloud.android.lib.resources.webfinger.services.WebFingerService
+
+class OCRemoteWebFingerDatasource(
+    private val webfingerService: WebFingerService,
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt:29]:
```suggestion
    private val webFingerService: WebFingerService,
```

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            webfingerService.getInstancesFromWebfinger(
+                lookupServer = lookupServer,
+                rel = rel.uri,
+                username = username,
+                client = ownCloudClient
+            )
+        }
+    }
+
+    override fun getInstancesFromAuthenticatedWebfinger(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt:50]:
```suggestion
    override fun getInstancesFromAuthenticatedWebFinger(
```

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/RemoteWebFingerDatasource.kt [hunk 7]
```diff
@@ -17,6 +17,19 @@
...
+import com.owncloud.android.domain.webfinger.model.WebFingerRel
+
+interface RemoteWebFingerDatasource {
+    fun getInstancesFromWebFinger(
+        lookupServer: String,
+        rel: WebFingerRel,
+        username: String
+    ): List<String>
+
+    fun getInstancesFromAuthenticatedWebfinger(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/RemoteWebFingerDatasource.kt:29]:
```suggestion
    fun getInstancesFromAuthenticatedWebFinger(
```
This should be renamed as well 👍

File: owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    override fun getInstancesFromWebFinger(
+        lookupServer: String,
+        rel: WebFingerRel,
+        username: String
+    ): List<String> {
+        val ownCloudClient = clientManager.getClientForAnonymousCredentials(lookupServer, false)
+
+        return executeRemoteOperation {
+            webfingerService.getInstancesFromWebfinger(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/webfinger/datasources/implementation/OCRemoteWebFingerDatasource.kt:41]: In the library as well 😀 
```suggestion
            webfingerService.getInstancesFromWebFinger(
```

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some more renamings and we're done @abelgardep 👍

**Review by JuancaG05 — APPROVED** [review]
Ready for QA 🚀

**Review by butonic — APPROVED** [review]

**Comment by jesmrec** [comment]
QA checks

different authentication methods and available idPs

- [X] Basic auth
- [X] OAuth2
- [X] OIDC
- [ ] Webfinger v1 (enterprise)
- [X] oCIS traefik (Webfinger v2)
- [X] oCIS Wopi
- [X] oCIS Keycloak

**Comment by jesmrec** [comment]
All existing auth methods are working in this PR. New authenticated webfinger works, but current instances have same idP and lookup server. So, this is a first working approach but not the ultimate one. Take this as "testable" version of the feature. 

From my side, it can be merged.

**Pull request merged**